puzzle.js

```diff
@@ -9,32 +9,68 @@ class Puzzle {
     this.target = target;
     // Use a roundtrip via JSON to clone the target array
     this.state = JSON.parse(JSON.stringify(target));
+    this.rows = this.state.length;
+    this.cols = this.state[0].length;
     this.difficulty = difficulty;
   }
 
-  shuffle(n) {
-    // if n>100, random moves should be enough to get an interesting start
-    // position. The advantage of using random moves is that it guarantees that
-    // the resulting board is solvable. The downside is that not all board
-    // positions will be equally likely to show up.
-    for (let i = 0; i < n; i++) {
-      // pick a number between 0-3
-      const pos = (Math.random() * 4) | 0;
-
-      // pick -1 or 1
-      const dir = ((Math.random() * 2) | 0) * 2 - 1;
-
-      if (Math.random() > 0.5) {
-        this.moveHorz(pos, dir);
-      } else {
-        this.moveVert(pos, dir);
+  shuffle() {
+    // special case for trivial puzzles
+    if (this.rows == 1) {
+      const n = randInt(this.cols - 1) + 1;
+      this.moveHorz(0, n);
+      return;
+    } else if (this.cols == 1) {
+      const n = randInt(this.rows - 1) + 1;
+      this.moveVert(0, n);
+      return;
+    }
+    const to2d = (n) => {
+      const j = n % this.cols;
+      const i = (n - j) / this.cols;
+      return [i, j];
+    };
+    const get = (n) => {
+      const [i, j] = to2d(n);
+      return this.state[i][j];
+    };
+    const set = (n, x) => {
+      const [i, j] = to2d(n);
+      this.state[i][j] = x;
+    };
+    const swap = (n1, n2) => {
+      const x = get(n1);
+      set(n1, get(n2));
+      set(n2, x);
+    };
+    // we generate a random permutation of all the cells, by swapping
+    // the first cell with a random one, then the second with a random
+    // one etc... And we fix parity at the end if necessary
+    const n = this.rows * this.cols;
+    let odd = false;
+    for (let n1 = 0; n1 < n - 1; n1++) {
+      const n2 = randInt(n - n1) + n1;
+      if (n1 != n2) {
+        odd = !odd;
+        swap(n1, n2);
       }
     }
+    const puzzle_odd = this.isOdd();
+    if (odd && !puzzle_odd) {
+      swap(0, 1);
+    }
+  }
+
+  // checks if odd permutation are possible for that puzzle. It's used
+  // to fix the state reached when doing a permutation of the puzzle
+  // during shuffle
+  isOdd() {
```

Comment:
It seems error prone to generate a puzzle with random swaps and then fix it in some cases. What's the advantage over just making a large number of random oves?

Reply:
It was more for the sport than anything else. The two (ridiculous) advantages are performance and choosing uniformly a shuffle. By the way, the actual algorithm can reach only 50% of permutations for odd puzzles because it always do an even number of permutations. Making an additional action on a coin toss would solve that little issue.

Reply:
for the existing easy and medium puzzles, all permutations are reachable. For the existing hard puzzle, only half are reachable. I'd rather we keep the existing shuffle mechanism, speed isn't an issue.

Reply:
For the easy and medium, yes all positions are reachable. But doing 100 moves (or any even number of moves) will never reach the "odd" permutations.